persistence/sql/persister_session.go

```diff
@@ -76,21 +90,50 @@ func (p *Persister) ListSessionsByIdentity(ctx context.Context, iID uuid.UUID, a
 	return s, nil
 }
 
+// UpsertSession creates a session if not found else updates.
+// This operation also inserts Session device records when a session is being created.
+// The update operation skips updating Session device records since only one record would need to be updated in this case.
 func (p *Persister) UpsertSession(ctx context.Context, s *session.Session) error {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UpsertSession")
 	defer span.End()
 
 	s.NID = p.NetworkID(ctx)
 
-	if err := p.Connection(ctx).Find(new(session.Session), s.ID); errors.Is(err, sql.ErrNoRows) {
-		// This must not be eager or identities will be created / updated
-		return errors.WithStack(p.GetConnection(ctx).Create(s))
-	} else if err != nil {
-		return errors.WithStack(err)
-	}
+	return errors.WithStack(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) error {
+		if err := sqlcon.HandleError(tx.Where("id = ? AND nid = ?", s.ID, s.NID).First(new(session.Session))); errors.Is(err, sqlcon.ErrNoRows) {
```

Comment:
Maybe instead, we could do this. I know that it's hard to review this snippet, but maybe @kelkarajay can work with this? I find this a lot more readable. Sorry, for making such a fuzz about this, but as a relative newcomer to the codebase, I find the persister code hard to read as is, because of all the pop magic, and making it clearer is a win, right?

```go

	return errors.WithStack(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) error {

		exists, err := tx.Where("id = ? AND nid = ?", s.ID, s.NID).Exists(new(session.Session))
		if err != nil {
			return sqlcon.HandleError(err)
		}

		if exists {
			// This must not be eager or identities will be created / updated
			// Only update session and not corresponding session device records
			return sqlcon.HandleError(tx.Update(s)) // should be correct?
		}

		// This must not be eager or identities will be created / updated
		if err := sqlcon.HandleError(tx.Create(s)); err != nil {
			return err
		}

		for i := range s.Devices {
			device := &(s.Devices[i])
			device.SessionID = s.ID
			device.NID = s.NID

			if device.Location != nil {
				device.Location = stringsx.GetPointer(stringsx.TruncateByteLen(*device.Location, SessionDeviceLocationMaxLength))
			}
			if device.UserAgent != nil {
				device.UserAgent = stringsx.GetPointer(stringsx.TruncateByteLen(*device.UserAgent, SessionDeviceUserAgentMaxLength))
			}

			if err := sqlcon.HandleError(tx.Create(device)); err != nil {
				return err
			}

			s.Devices[i] = *device
		}
		return nil

	}))
```

Reply:
It would have been nice to have the insert function like https://gorm.io/docs/create.html#Upsert-x2F-On-Conflict. Worth considering upstream contribution to pop, yeah. In the interim, the snippet above looks good tbh

Reply:
Went with this ^